docs/root/overview.md

```diff
@@ -1,60 +1,70 @@
-gRPC and/or CLI.
-
-### Streaming parameterization and output stats
-
-Once we have phases, the gRPC service, and perhaps the CLI, would be natural
-candidates to follow up with to allow dynamic phase injection, as well as send
-back reports per phase.
+## Architecture
+
+The following diagram outlines the overall architecture of Nighthawk.
+
+![Nighthawk architecture](images/nighthawk_architecture.png)
+
+Nighthawk exposes its API either via a [command line
+interface](../../source/exe/client_main_entry.cc) (CLI) or as a [gRPC
+server](../../source/exe/service_main_entry.cc) with a [proto
+API](../../api/client/service.proto).
+When running as a CLI, Nighthawk will run a single execution and print out the
+results. When running as a gRPC server, Nighthawk is capable of performing
+multiple executions one after another, returning results to the gRPC client.
+
+In either of the execution modes, the
+[Options](../../source/client/options_impl.h) object abstracts all the
+[options](../../api/client/options.proto) selected by the user and provides them
+to the rest of the Nighthawk architecture.
+
+Nighthawk creates a [Process](../../source/client/process_impl.h)
+instance that generates Envoy bootstrap configuration capable of upstreaming
+requests with the protocol and options selected by the user. This bootstrap is
+used to initialize an Envoy upstream cluster manager.
+
+The **Process** instance creates multiple [Client
+workers](../../source/client/client_worker_impl.h) that get access to the Envoy
+upstream cluster manager and drive the execution.
+See the [next section](#client-worker-architecture) for a detail view into the
+worker architecture.
+
+Once the execution finishes, the [Output
+Collector](../../source/client/output_collector_impl.h)
+provides results collected from individual **Client workers** to the [Output
+formatter](../../source/client/output_formatter_impl.h)
+which represents the results in the format selected by the user.
+
+### Client worker architecture
+
+![Client worker architecture](images/client_worker_architecture.png)
+
+Each **Client worker** runs a
+[Sequencer](../../source/common/sequencer_impl.h)
+which is responsible for maintaining the pace of requests that was specified by
+the user. To maintain the pace, the **Sequencer** queries a
+[RateLimiter](../../source/common/rate_limiter_impl.h)
+for [request-release timing](terminology.md#request-release-timing). When it
+is time to release a request, the
+[BenchmarkClient](../../source/client/benchmark_client_impl.h)
+is requested to do so by the **Sequencer**.
+
+**BenchMarkClient** retrieves data for the next request from the configured
+[RequestSource](../../source/common/request_source_impl.h)
+and asks its underlying **Connection pool** to create a
+[StreamDecoder](../../source/client/stream_decoder.h)
+which releases the request. The **StreamDecoder** sends the request and emits
+events as it progresses (connection pool ready, completion, etc...).
```

Comment:
It's still difficult to tell what the Stream Decoder is doing here.  In particular, we are saying the stream decoder "sends the request and emits evens as it progress" which begs the question - what about that is decoding?

Also what are the events, which aren't in the diagram?

Reply:
Moved the examples of events (the parenthesis) right after the "event" word. Also clarified the role of StreamDecoder.

Done in 7ba9de7ba336819db5977c8876e69084abe35ea1.